native/membrane_videocompositor/src/elixir_bridge/elixir_structs.rs

```diff
@@ -136,3 +140,710 @@ impl From<ElixirCropping> for Box<dyn TextureTransformation> {
+        let final_object = &objects[&self.output];
+
+        fn parse_object(
+            name: &ObjectName,
+            object: &Object,
+            nodes: &mut HashMap<ObjectName, Arc<Node>>,
+            objects: &HashMap<ObjectName, Object>,
+        ) -> Result<Arc<Node>, SceneParsingError> {
+            if let Some(node) = nodes.get(name) {
+                return Ok(node.clone());
+            }
+
+            match object {
+                Object::Video(InputVideo { input_pad }) => {
+                    let node = Arc::new(Node::Video {
+                        pad: input_pad.clone(),
+                    });
+                    nodes.insert(name.clone(), node.clone());
+                    Ok(node)
+                }
+
+                Object::Texture(Texture {
+                    input,
+                    transformations,
+                    resolution,
+                }) => {
+                    let mut current = parse_object(input, &objects[input], nodes, objects)?;
+
+                    // we may need to get ownership of the transformation here later on in the project.
+                    // this means that we'll most likely need to remove `object` from `objects`, but since it's not certain
+                    // I'm leaving it as is for now
+                    for &transformation in transformations {
+                        current = Arc::new(Node::Transformation {
+                            previous: current,
+                            transformation,
+                            resolution: resolution.clone(),
+                        });
+                    }
+
+                    nodes.insert(name.clone(), current.clone());
+
+                    Ok(current)
+                }
+
+                Object::Layout(Layout {
+                    inputs,
+                    resolution,
+                    implementation,
+                }) => {
+                    let inputs = inputs
+                        .iter()
+                        .map(|(internal_name, object_name)| {
+                            parse_object(
+                                object_name,
+                                objects.get(object_name).ok_or_else(|| {
+                                    SceneParsingError::UndefinedName(object_name.clone())
+                                })?,
+                                nodes,
+                                objects,
+                            )
+                            .map(|node| (internal_name.clone(), node))
+                        })
+                        .collect::<Result<HashMap<_, _>, _>>()?;
+
+                    let result = Arc::new(Node::Layout {
+                        resolution: resolution.clone(),
+                        implementation: *implementation,
+                        inputs,
+                    });
+                    nodes.insert(name.clone(), result.clone());
+                    Ok(result)
+                }
```

Comment:
With separate functions for that logic, it would be more readable.

Reply:
I actually think it would be less readable, since the code for recursing into different variants would be in separate places in the file. I find it nicer to read like it is now.

Reply:
notice how clippy doens't mark it as bad. this level of indentation is rather normal in conventional rust.

Reply:
It is a different language, but the 60 lines long match is unreadable. There are a few problems I'm noticing here:
1. This `parse_object` function is defined in `Scene` when it depends on the `Object` structure. This violates the single responsibility principle. When we change the definition of `Object`, we should expect only to change the `Object` implementation, but here we have actually to change the `Scene` struct. This creates unnecessary dependency, complicates the codebase, and makes it harder to extend in the future. I would move it to `Object` `impl`
2. There is an inconsistency between this giant pattern match, which handles the whole logic for parsing all types of objects, and the `mantioned_names` pattern match, which does it with separate functions. This pattern match will even grow bigger once we add new objects / change the current one.
3. Chandling those conversions in a single place, means that we have to modify it whenever we change the single one. By splitting this logic to separate functions for each type of `Object`, we only have to modify the code responsible for the appropriate `Object` implementation.

Therefore, I think that:
1. `parse_object` should be moved to `Object`
2. Each object type should handle its parsing inside their own implementation

Reply:
3. Each of these match arms repeats this logic:
```rust
                    nodes.insert(name.clone(), current.clone());
                    Ok(node)
```

Reply:
How does it make it harder to extend in the future? any change in the definition of any of these structs will force you (by erroring out during compilation) to change the code here as well. This will happen automatically, you will be brought to this code by the compiler. The definitions are not in some remote place at the other end of the project. They're in the same module.

I'd say the SRP in rust applies more to modules than to individual structs. Also, quite often data-oriented design is preferred over object oriented programming patterns in rust. This means that you treat individual structs more as data which is open to being modified and processed and less as self-contained objects that talk to each other.

The inconsistency between this function and `mentioned_names` stems from the fact that the functionality in `mentioned_names` could be used somewhere else in the module (I think this even is the case), and can be derived from the object alone, without any external context. On the other hand, I cannot come up with a way to utilize `parse_object` in a different part of this module and parsing doesn't really make sense to be done without the additional information contained in the scene. I believe (at least from the data-oriented design point of view) since all of the code does a very specific thing, which is almost certainly useless for the rest of the project, it makes sense to put all of it in one spot where it is used.

The third point doesn't make sense for me. Yes, we have to modify this code whenever one of these objects changes. I don't see why it would be nicer to edit the code for parsing `Texture` in `Texture::parse` rather the part of the match responsible for handling `Texture`. There is no difference, you only have to edit the code responsible for `Texture` in both cases.

Last part: I opened the repo for the rustc compiler, which is, to my knowledge, the largest rust project in existence. Hundreds of people work on it together daily. looked through <10 files and found [this](https://github.com/rust-lang/rust/blob/e49122fb1ca87a6c3e3c22abb315fc75cfe8daed/compiler/rustc_ast_passes/src/ast_validation.rs#L806). I know this is a different case than what we have here, I just wanted to demonstrate that very long match statements matching over enums being located outside of these enums' `impl` blocks are considered rather normal in rust. I wouldn't call a 60 lines long match statement 'unreadable', since these are commonplace even in extremely large projects, which tend to place strong emphasis on the code being readable to all contributors.

I will move this match to `Object` however, because I don't think the negative consequences of doing that are significant enough to discuss further.